### ISSUE TITLE
New movement code bugs

### ISSUE THREAD

**Issue by 12Me21** [minetest/minetest#7725]
##### Issue type
 - Bug report
 
##### Minetest version
```
5.0.0-a4f41e7
0.4.17.1
```

##### OS / Hardware

Operating system: Windows 7, Windows 10 (64 bit)
CPU: (various)

##### Steps to reproduce
For this, it is very helpful to have a mod like technic or elepower, as the power cables cause many movement glitches. I'll be using Elepower power conduits here, but technic LV cables should work just as well.

1: Sneak upwarp:
Build a 2 node high wall and place a cable next to the bottom of the wall.
Stand on the cable, move up against the wall, and sneak+jump.
You should "snap" upwards to the top surface of the wall.

1.5: Can't jump!
Now, without moving, release the sneak button and try to jump again from your new position.

2: Sneak upwarp 2:
![image](https://user-images.githubusercontent.com/4053256/45469899-7b818d80-b6f9-11e8-9e1f-adcb5b15b18b.png)
Build a vertical square of cables like this.
Then stand next to it and press sneak and jump. You'll be teleported to the top surface.

1.5.5: Can't jump, part 2:
This is a bit tricky. From your previous position on top of the cables, hold sneak and move backwards as far as you can. Then release the sneak button and carefully move sideways so you're at the middle:
![image](https://user-images.githubusercontent.com/4053256/45471343-ceaa0f00-b6fe-11e8-9c78-aa1c38ec0e59.png)
Now, you won't be able to jump.

3: Getting stuck sneaking on a flat surface:
Now, on the same structure, try to walk from one end to the other, while sneaking. You will not be able to cross the middle point.

3.5: Sneaking is broken, ok?
Without sneaking, move to the middle of that surface. Pressing the sneak key will teleport you sideways slightly.

4: Difficult to reproduce:
In rare cases, it is possible to walk *through* blocks. I haven't managed to find a repeatable setup for this yet, but like the other glitches, it seems to happen a lot around cables.

All of these issues appear to be solved if you use `set_physics_override` with `new_move` disabled.
Of course, the new movement code was introduced to fix the old sneak glitch, but I think the sneak glitch is a lot less glitchy than what we have now...
I believe some of the problems are related to connected nodeboxes, but definitely not all of them.

Anyway, sorry this issue is so long and disorganized. As I was coming up with situations to demonstrate the problem, I kept finding more and more bugs...

TL;DR:
The new movement code has some bugs, and there are situations where you are standing on the ground but aren't able to jump.

**Comment by paramat** [comment]
Please could you paste the node definition of the Elepower power conduit here?
Certainly new move has some bugs, just as old move does.
@SmallJoker 
Related #5438 #5535

**Comment by 12Me21** [comment]
```Lua
node_box = {
	type = "connected",
	fixed = {
		{-cd, -cd, -cd, cd, cd, cd}
	},
	connect_front = {
		{-cd, -cd, -1/2, cd, cd, -cd}
	},
	connect_back = {
		{-cd, -cd, cd, cd, cd, 1/2}
	},
	connect_top = {
		{-cd, cd, -cd, cd, 1/2, cd}
	},
	connect_bottom = {
		{-cd, -1/2, -cd, cd, -cd, cd}
	},
	connect_left = {
		{-1/2, -cd, -cd, cd, cd, cd}
	},
	connect_right = {
		{cd, -cd, -cd, 1/2, cd, cd}
	},
},
```
(`cd` is `1/8`)

**Comment by sfan5** [comment]
@SmallJoker

**Comment by SmallJoker** [comment]
Oh man. Yet another issue for your sneak-related bugs list @Fixer-007 

>1: Sneak upwarp:

So jumping from a cable onto a wall, in total 1.5m jump height? Doesn't sound like a bug.
>1.5: Can't jump!

Not a bug. You can also not jump when standing on an outer edge of any regular node.
> 2: Sneak upwarp 2:

If you can reproduce this with `sneak_glitch` (physics override) disabled, then it's a bug.
>1.5.5: Can't jump, part 2:

Same as 1.5. Not a bug.
>3: Getting stuck sneaking on a flat surface:

So it has troubles finding the next best node to sneak. Should work closer to the node center though. Will have to test this.
>3.5: Sneaking is broken, ok?

Same cause as 3, teleporting is caused because Minetest thinks that you might fall off the sneaking node.
>4: Difficult to reproduce: (glitching)

Also happens without sneaking or anything. Collision detection issue.
>Also, I hope this doesn't stir up any old controversy... I'm not suggesting that the new movement code be removed, or the old sneak behavior be enabled by default. This is just about bugs and problems with the new movement code.

The entire sneaking code is a minefield. Touch one line and arise two new issues. This issue has potential to start one of those long discussions from scratch again... well great.

**Comment by SmallJoker** [comment]
> Not being able to jump while standing on the ground is most certainly a bug.
> 1.5: Can't jump!
> 1.5.5: Can't jump, part 2:

This would arise more issues (such as #3767). If you'd like to test this yourself, [remove or tweak `d` and step up on stairs](https://github.com/minetest/minetest/blob/81c06dea9b6b863e0c8f6a9c052a5cea3a130eab/src/collision.cpp#L572).
Meanwhile it works in the sneak code because it co-exists with the collision engine - which again causes side-effects.
I did some research on the reported bugs and things turned out to be very simple.
> 1: Sneak upwarp:

A simple sneak-jump. The old sneak code allowed 2-node high jumps sometimes.
> 3: Getting stuck sneaking on a flat surface:
> 3.5: Sneaking is broken, ok?
The collision engine handles connected nodeboxes; sneak doesn't. The old code handled everything as a 1m * 1m * {variable height} node, so it didn't matter how small the collisionbox actually was. Since the new code ensures that you won't fall off a regular node while sneaking, this nodebox calculation would have to be adapted to the sneak code as well (or import the data from the collision result).
Teleporting happens again towards the "fixed" nodebox definition, since the connections are not calculated.

**Comment by paramat** [comment]
12Me12 what's your setting for 'sneak_glitch'?
https://github.com/minetest/minetest/blob/766fb7b46edab47f96b8b940e390daf5319261b7/doc/lua_api.txt#L5150

**Comment by 12Me21** [comment]
Yeah I have `sneak_glitch` disabled.

Anyway, I think these are some reasonable expectations for movement, which minetest fails at:
1. If you are standing on the ground, you should be able to jump
2. Sneaking should prevent you from falling off ledges
3. While sneaking, you should be able to walk around on flat surfaces without being obstructed
4. Sneaking should not allow you to stand in places you normally wouldn't be able to (as in the old movement code).

**Comment by paramat** [comment]
See #5861

**Comment by paramat** [comment]
#7727 merged.

**Comment by Zughy** [comment]
Make little sense to keep an issue about multiple sneak bugs reported at once, closing
